src/slice.rs

```diff
@@ -79,6 +80,18 @@ macro_rules! impl_slice_from_index_type {
             }
         }
 
+        impl From<RangeInclusive<$index>> for Slice {
+            #[inline]
+            fn from(r: RangeInclusive<$index>) -> Slice {
+                let end = *r.end() as isize;
+                Slice {
+                    start: *r.start() as isize,
+                    end: if end == -1 { None } else { Some(end + 1) },
+                    step: 1,
+                }
+            }
+        }
```

Comment:
Line 70 to 129 and 265 to 324. do you think it would a good place for a macro? They are already macros, but it could be a merge of `impl_slice_from_index_type` and `impl_sliceorindex_from_index_type`. The only change in all those lines is `Slice` and `SliceOrIndex`.

It's a real question btw. Maybe too much macro make the code unreadable or un-fun or wathever. Or maybe the special case `From<$index>` make this more complex?

Reply:
I've merged the macros as you've suggested, which substantially reduced the duplication. The special case is not an issue; I've kept a separate macro for it.

Thanks for pointing this out!